Add initial spec for auth service

mark-burnett opened the pull request:
This is a draft based on some combined ideas from the OAuth toy code bases that have been used for experimentation:  https://github.com/mark-burnett/oauth-skeleton and https://github.com/iferguson90/oauth-toy.

It is not complete and may require significant changes to get it closer to what we want.

mark-burnett committed:
add auth-related RFCs

mark-burnett committed:
add initial auth server spec

mark-burnett committed:
add quotes to reference descriptions

mark-burnett committed:
auth: clarify introduction of GET /authorization

mark-burnett committed:
auth: elaborate 403 error causes in /validate

mark-burnett committed:
auth: remove requirement registering live clients

mark-burnett committed:
auth: fix WWW-Authenticate header for /validate

mark-burnett committed:
add auth description to README

mark-burnett committed:
auth: add query endpoint for user roles

mark-burnett committed:
auth: add /represents endpoint

mark-burnett committed:
auth: only use 403 status to mean bad Auth header

mark-burnett committed:
auth:  whitespace

mark-burnett committed:
auth: missing header

mark-burnett committed:
auth: add details for POST /clients

davidlmorton commented on auth.md:
```diff
@@ -0,0 +1,365 @@
+
+Error:
+
+- HTTP 302 (Found)
```
Is 302 (found) an error?

mark-burnett replied:
Well, the OAuth spec says that "the authorization server informs the client" of the error by adding additional information into the query string of the `redirect_uri` (section 4.1.2.1), so it's one of the status codes that should be raised for certain kinds of errors.

mark-burnett committed:
auth:  allow non-confidential clients

mark-burnett committed:
auth: remove references to PAM

We might use PAM, or we might just use POSIX calls.

mark-burnett committed:
auth: clarify parameters for POST /tokens

mark-burnett committed:
auth: clarify generated fields in POST /clients

mark-burnett committed:
auth: add section for resource server APIs

mark-burnett committed:
auth: whitespace

mark-burnett committed:
auth: remove question of special users' API keys

mark-burnett committed:
auth: add initial justification for service

mark-burnett committed:
auth: add /v1 to non-OAuth API paths

mark-burnett committed:
auth:  Authorization API is for resource servers

Add separate handling for resource servers and clients, so that we can
clearly differentiate them.

iferguson90 committed:
Update auth.md

Clarify authN and authZ

iferguson90 committed:
Update auth.md

Clarify justification

iferguson90 committed:
Remove optimization note

Implementation details are not necessary for the specification

iferguson90 committed:
Update auth.md

redirect_uri is not required for implicit grant

iferguson90 committed:
Clarify response_type

iferguson90 committed:
Note that `state` is recommended in RFC

iferguson90 commented on auth.md:
```diff
@@ -0,0 +1,371 @@
+    - This indicates that the user agent should retry the request with
+      credentials.
+    - Includes the header `WWW-Authenticate: API-Key`.
+    - Includes the header `WWW-Authenticate: Basic`.
```
Do we want to support basic HTTP authentication for the authorization endpoint?

iferguson90 committed:
Fix repeated word

iferguson90 committed:
Clarify `redirect_uri` in POST /tokens

iferguson90 committed:
Clarify `refresh_token` requirements

iferguson90 committed:
Break out params based on grant_type for POST /tokens

iferguson90 committed:
Consistency

iferguson90 committed:
Remove /v1 for consistency

iferguson90 committed:
Clarify and reword POST /validate

iferguson90 committed:
Clarify POST /validate responses

iferguson90 committed:
Remove /v1 from all URLs

We can add them back later

iferguson90 committed:
NO TRAILING WHITESPACE

iferguson90 committed:
Consolidate errors for POST /validate

iferguson90 committed:
Remove POST /represents

iferguson90 committed:
Work on POST /clients

iferguson90 committed:
Remove extra "Error:" line

mkiwala committed:
Remove Resource Server API section

mkiwala committed:
Remove Authorization API section

mkiwala committed:
Remove specification duplicated from OAuth 2.0 RFC.

mkiwala committed:
Further editing of /token endpoint section

mkiwala committed:
Fixup /authorization endpoint

mkiwala committed:
Add an ID Tokens section

mark-burnett committed:
whitespace

mark-burnett committed:
add new thoughts about callbacks

mark-burnett committed:
auth: replace PUT with PATCH

mark-burnett committed:
auth: add /v1 to endpoint paths

mark-burnett committed:
auth: update properties for creating clients

mark-burnett committed:
auth: update some authentication requirements
